lib/fencing/st_rhcs.c

```diff
@@ -30,37 +32,55 @@ stonith__list_rhcs_agents(stonith_key_value_t **devices)
 {
     // Essentially: ls -1 @sbin_dir@/fence_*
 
-    int count = 0;
+    int count = 0, i;
     struct dirent **namelist;
-    int file_num = scandir(RH_STONITH_DIR, &namelist, 0, alphasort);
-
-    if (file_num > 0) {
+    const int file_num = scandir(RH_STONITH_DIR, &namelist, 0, alphasort);
+
+#if (!defined(_POSIX_C_SOURCE) || _POSIX_C_SOURCE < 200809L) \
+        && (!defined(O_SEARCH) || !defined(O_PATH))
+    char buffer[FILENAME_MAX + 1];
+#elif defined(O_SEARCH)
+    const int dirfd = open(RH_STONITH_DIR, O_SEARCH);
+#else
+    const int dirfd = open(RH_STONITH_DIR, O_PATH);
+#endif
+
+    for (i = 0; i < file_num; i++) {
         struct stat prop;
-        char buffer[FILENAME_MAX + 1];
 
-        while (file_num--) {
-            if ('.' == namelist[file_num]->d_name[0]) {
-                free(namelist[file_num]);
-                continue;
-
-            } else if (!crm_starts_with(namelist[file_num]->d_name,
-                                        RH_STONITH_PREFIX)) {
-                free(namelist[file_num]);
+        if (crm_starts_with(namelist[i]->d_name, RH_STONITH_PREFIX)) {
+#if !defined(_POSIX_C_SOURCE) || _POSIX_C_SOURCE < 200809L
```

Comment:
We need the exact condition as before here (and at the end of the function), otherwise we get into the wrong block (e.g. posix not defined but O_SEARCH supported). It would be a good idea to define a constant for the condition.